### PR TITLE
5.x query abstract

### PR DESCRIPTION
**Pull request by ADmad**
- Make `Datbase\Query` abstract
- Code cleanup

Refs #16648

### CODE REVIEW

File: src/Database/Connection.php [hunk 1]
```diff
@@ -278,16 +278,6 @@ public function run(Query $query): StatementInterface
...
 
-    /**
-     * Create a new Query instance for this connection.
-     *
-     * @return \Cake\Database\Query
-     */
-    public function newQuery(): Query
-    {
-        return new Query($this);
-    }
```
Comment (othercorey) [inline src/Database/Connection.php:289]: I wonder how big if a break this will be. There's nothing we could transform calls to newQuery() into.

File: src/Database/IdentifierQuoter.php [hunk 2]
```diff
@@ -59,12 +60,14 @@ public function quote(Query $query): Query
...
+            $this->_quoteSelect($query);
+        } elseif ($query instanceof InsertQuery) {
             $this->_quoteInsert($query);
         } elseif ($query instanceof UpdateQuery) {
             $this->_quoteUpdate($query);
-        } else {
-            $this->_quoteParts($query);
+        } elseif ($query instanceof DeleteQuery) {
+            $this->_quoteDelete($query);
         }
```
Comment (othercorey) [inline src/Database/IdentifierQuoter.php]: should this have an else throw?

File: src/Database/IdentifierQuoter.php [hunk 3]
```diff
@@ -101,19 +104,16 @@ public function quoteExpression(ExpressionInterface $expression): void
...
     /**
-     * Quotes all identifiers in each of the clauses of a query
+     * Quotes all identifiers in each of the clauses/parts of a query
      *
      * @param \Cake\Database\Query $query The query to quote.
+     * @param array $parts Query clauses.
      * @return void
      */
-    protected function _quoteParts(Query $query): void
+    protected function _quoteParts(Query $query, array $parts): void
```
Comment (othercorey) [inline src/Database/IdentifierQuoter.php:116]: Having the parts list separate from the Query instance doesn't feel right. When would we describe a query different than itself?

Comment (ADmad) [reply src/Database/IdentifierQuoter.php:116]: These are just the part/clause names which need to be processed. They are different depending on the query type. The actual values for these parts are fetched from the Query itself.

File: src/Database/IdentifierQuoter.php (hunk 2 above)
Comment (ADmad) [reply src/Database/IdentifierQuoter.php]: Guess we could for completeness.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -114,11 +114,11 @@ jobs:
 
         if [[ ${{ matrix.php-version }} == '8.1' ]]; then
           export CODECOVERAGE=1
-          vendor/bin/phpunit --verbose --coverage-clover=coverage.xml
-          CAKE_TEST_AUTOQUOTE=1 vendor/bin/phpunit --verbose --testsuite=database
+          vendor/bin/phpunit tests/TestCase/Database --verbose --coverage-clover=coverage.xml
+          CAKE_TEST_AUTOQUOTE=1 vendor/bin/phpunit tests/TestCase/Database --verbose --testsuite=database
```
Comment (markstory) [inline .github/workflows/ci.yml:118]: Do we need to revert this change before merging?

File: src/Database/IdentifierQuoter.php (hunk 2 above)
Comment (markstory) [reply src/Database/IdentifierQuoter.php]: Could get fancy and use `match`.

File: src/Database/Connection.php (hunk 1 above)
Comment (markstory) [reply src/Database/Connection.php:289]: We *could* backport the `newSelectQuery()` methods to 4.x and add a deprecation to `newQuery()`. That would let us encourage users upgrade in a reasonable timeframe.

I'm happy to take care of that when this merges.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (ADmad) [reply .github/workflows/ci.yml:118]: Not yet, I will revert it when I getting to updating the ORM once the changes to database package are done.

File: src/Database/IdentifierQuoter.php (hunk 2 above)
Comment (ADmad) [reply src/Database/IdentifierQuoter.php]: Right, can use `match()`, still haven't got accustomed to the fact that it can use expressions for matching :).

Comment (ADmad) [reply src/Database/IdentifierQuoter.php]: Done

**Comment by ADmad** [comment]
Anymore feedback on this?
